### PR TITLE
[ELASTIC-86] Implement rolling upgrades

### PR DESCRIPTION
**Pull request by loren**

### CODE REVIEW

**Comment by gabrielhartmann** [comment]
@sascala: Can you have a look at the upgrade directions please?

File: frameworks/elastic/src/main/dist/svc.yml [hunk 1]
```diff
@@ -262,15 +262,45 @@ plans:
...
+      data:
+        strategy: {{UPDATE_STRATEGY}}
+        pod: data
+      ingest:
+        strategy: {{UPDATE_STRATEGY}}
+        pod: ingest
+      coordinator:
+        strategy: {{UPDATE_STRATEGY}}
+        pod: coordinator
+  upgrade:
```
Comment (gabrielhartmann) [inline frameworks/elastic/src/main/dist/svc.yml]: There's no such thing as an `upgrade` plan yet.  I don't see it added in this PR.  Did I miss something?

File: frameworks/elastic/tests/test_upgrade.py [hunk 2]
```diff
@@ -3,12 +3,13 @@
...
+
 @pytest.mark.upgrade
 @pytest.mark.sanity
 def test_upgrade_downgrade():
     sdk_test_upgrade.upgrade_downgrade(
         "beta-{}".format(PACKAGE_NAME),
         PACKAGE_NAME,
         DEFAULT_TASK_COUNT,
-        additional_options={ "service": {"beta-optin": True} },
+        additional_options={"service": {"beta-optin": True, "upgrade_strategy": "parallel"}},
```
Comment (gabrielhartmann) [inline frameworks/elastic/tests/test_upgrade.py]: `upgrade_strategy` should be removed when the `upgrade` plan is removed

File: frameworks/elastic/universe/config.json [hunk 3]
```diff
@@ -34,12 +34,24 @@
           "description": "Elasticsearch deploy strategy. [parallel, serial]",
           "type": "string",
           "default": "parallel"
+        },
+        "update_strategy": {
+          "description": "Elasticsearch update strategy. [parallel, serial, safe]",
+          "type": "string",
+          "default": "serial"
+        },
+        "upgrade_strategy": {
```
Comment (gabrielhartmann) [inline frameworks/elastic/universe/config.json]: This section should be removed.

File: frameworks/elastic/universe/config.json [hunk 4]
```diff
@@ -34,12 +34,24 @@
...
+          "default": "safe"
         }
       },
       "required": [
         "name",
         "user",
-        "deploy_strategy"
+        "deploy_strategy",
+        "update_strategy",
+        "upgrade_strategy"
```
Comment (gabrielhartmann) [inline frameworks/elastic/universe/config.json]: This element should be removed.

File: frameworks/elastic/universe/marathon.json.mustache [hunk 5]
```diff
@@ -46,6 +46,8 @@
     {{/service.virtual_network}}
 
     "DEPLOY_STRATEGY": "{{service.deploy_strategy}}",
+    "UPDATE_STRATEGY": "{{service.update_strategy}}",
+    "UPGRADE_STRATEGY": "{{service.upgrade_strategy}}",
```
Comment (gabrielhartmann) [inline frameworks/elastic/universe/marathon.json.mustache]: ditto

File: frameworks/helloworld/tests/test_safe_strategy.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            'world': {'count': 1}
+        },
+        check_suppression=False)
+
+
+def teardown_module(module):
+    install.uninstall(PACKAGE_NAME)
+
+
+@pytest.mark.safe
```
Comment (gabrielhartmann) [inline frameworks/helloworld/tests/test_safe_strategy.py]: Probably don't need this mark.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/strategy/SafeStrategy.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * A SafeStrategy requires that the user manually invoke {@link #proceed()} calls in order to move along a deployment.
+ * This allows the user to manually check that the rollout looks good at each Step.
+ * <p>
+ * The default is to do nothing until a first {@link #proceed()} call is received, at which point the first {@link Step}
+ * is deployed. The user may then examine whether the deployment of that first {@link Step} was successful before
+ * invoking the next {@link #proceed()} call, continuing until the deployment is complete.
+ */
+public class SafeStrategy extends CanaryStrategy {
+
+
```
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/strategy/SafeStrategy.java]: extra newline

**Comment by loren** [comment]
Thank you for the quick and thorough review @gabrielhartmann. All your comments have been addressed.

**Comment by loren** [comment]
retest this please

CCM launch failed

File: frameworks/elastic/universe/config.json [hunk 8]
```diff
@@ -34,12 +34,18 @@
           "description": "Elasticsearch deploy strategy. [parallel, serial]",
           "type": "string",
           "default": "parallel"
+        },
+        "update_strategy": {
+          "description": "Elasticsearch update strategy. [parallel, serial, safe]",
+          "type": "string",
```
Comment (vishnu2kmohan) [inline frameworks/elastic/universe/config.json]: `enum`?

Comment (loren) [inline frameworks/elastic/universe/config.json]: Good catch! Fixed.

**Comment by loren** [comment]
retest this please

(CCM failure)

File: frameworks/elastic/docs/upgrade.md [hunk 9]
```diff
@@ -21,3 +21,111 @@ $ dcos package install elastic --options=/path/to/options.json
 The command above will trigger the install of the new Elastic version. You can follow the upgrade progress by making a REST request identical to the one used to follow the progress of a configuration upgrade. See the Configuring section for more information.
 
 Note: The upgrade process will cause all of your Elastic node processes to restart.
+
+## Rolling Upgrades
+
+By default, the update strategy is `serial`, which waits for each node to be ready before moving on to the next one. But the elasticsearch node being ready doesn't mean all the shards in the indexes are ready. This may be an appropriate strategy for test clusters or clusters where writes can be suspended during the upgrade/update process, but for production clusters you may want to use the `safe` strategy.
```
Comment (sascala) [inline frameworks/elastic/docs/upgrade.md]: Should `elasticsearch` be capitalized?

Suggest:

The default upgrade strategy is `serial`. This strategy waits for each node to be ready before moving on to the next one. 

However, a node can be ready before all of the shards in the indexes are ready. For this reason, the `serial` upgrade strategy is only appropriate for test clusters or clusters where writes can be suspended during the upgrade/update process. For production clusters, use the `safe` strategy.

File: frameworks/elastic/docs/upgrade.md [hunk 10]
```diff
@@ -21,3 +21,111 @@ $ dcos package install elastic --options=/path/to/options.json
...
+
+In the `safe` strategy, the operator manually triggers the upgrade for each node, pausing to make sure the deployment is going as planned using `dcos elastic plan show`. The following rolling upgrade process is based on Elastic's own (recommendations for rolling upgrades)[https://www.elastic.co/guide/en/elasticsearch/reference/current/rolling-upgrades.html].
```
Comment (sascala) [inline frameworks/elastic/docs/upgrade.md]: link should be: `[recommendations for rolling upgrades](https://www.elastic.co/guide/en/elasticsearch/reference/current/rolling-upgrades.html)

File: frameworks/elastic/docs/upgrade.md [hunk 11]
```diff
@@ -21,3 +21,111 @@ $ dcos package install elastic --options=/path/to/options.json
...
 
 Note: The upgrade process will cause all of your Elastic node processes to restart.
+
+## Rolling Upgrades
+
+By default, the update strategy is `serial`, which waits for each node to be ready before moving on to the next one. But the elasticsearch node being ready doesn't mean all the shards in the indexes are ready. This may be an appropriate strategy for test clusters or clusters where writes can be suspended during the upgrade/update process, but for production clusters you may want to use the `safe` strategy.
+
+In the `safe` strategy, the operator manually triggers the upgrade for each node, pausing to make sure the deployment is going as planned using `dcos elastic plan show`. The following rolling upgrade process is based on Elastic's own (recommendations for rolling upgrades)[https://www.elastic.co/guide/en/elasticsearch/reference/current/rolling-upgrades.html]. 
+
+After starting the upgrade process, the plan will await your input to continue:
```
Comment (sascala) [inline frameworks/elastic/docs/upgrade.md]: "the plan will wait for your input to continue"

File: frameworks/elastic/docs/upgrade.md [hunk 12]
```diff
@@ -21,3 +21,111 @@ $ dcos package install elastic --options=/path/to/options.json
...
+│  ├─ data-0:[node] (WAITING)
+│  └─ data-1:[node] (WAITING)
+├─ ingest (WAITING)
+│  └─ ingest-0:[node] (WAITING)
+└─ coordinator (WAITING)
+   └─ coordinator-0:[node] (WAITING)
+```
+
+Steps 1 and 2 are the same as in Elastic's guide:
+1. Disable shard allocation
```
Comment (sascala) [inline frameworks/elastic/docs/upgrade.md]: - periods at the end of the list items.
- can number each `1.` and markdown will generate the correctly numbered list.

File: frameworks/elastic/docs/upgrade.md [hunk 13]
```diff
@@ -21,3 +21,111 @@ $ dcos package install elastic --options=/path/to/options.json
...
+   └─ coordinator-0:[node] (COMPLETE)
+```
+
+## Full Cluster Restarts
+
+Some changes to an Elasticsearch cluster require a full restart of all the nodes. A rolling upgrade is not possible under the following circumstances:
+  * Upgrading across major versions. Consult (this table)[https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-upgrade.html] to see if a full cluster restart is required. 
+  * Toggling X-Pack installation.
+  
+In these cases, you will need to use the `parallel` strategy and perform a full cluster restart using the following process based on Elastic's (full cluster restart guide)[https://www.elastic.co/guide/en/elasticsearch/reference/current/restart-upgrade.html]:
```
Comment (sascala) [inline frameworks/elastic/docs/upgrade.md]: same note about the links

**Comment by sascala** [comment]
@gabrielhartmann Done.

**Pull request closed without merging**